pyccl/ccl_background.i

```diff
@@ -5,103 +5,124 @@
 %}
 
 // Enable vectorised arguments for arrays
-%apply (double* IN_ARRAY1, int DIM1) {(double* a, int na)};
-%apply (double* IN_ARRAY1, int DIM1) {(double* chi, int nchi)};
-%apply (double* ARGOUT_ARRAY1, int DIM1) {(double* output, int nout)};
+%apply (double *IN_ARRAY1, int DIM1) {(double *a, int na)};
+%apply (double *IN_ARRAY1, int DIM1) {(double *chi, int nchi)};
+%apply (int DIM1, double *ARGOUT_ARRAY1) {(int nout, double *output)};
 
 %include "../include/ccl_background.h"
 
+/* The python code here will be executed before all of the functions that
+   follow this directive. */
+%feature("pythonprepend") %{
+    if numpy.shape(a) != (nout,):
+        raise CCLError("Input shape for `a` must match `(nout,)`!")
+%}
+
```

Comment:
This is an interesting hack.  It's not how I would have done it.  I would have added checks in the python code directly before passing into the C functions.  I'm not sure what all the implications are about doing it this way instead though.

The only possible downside I can think of is what the traceback looks like if a user hits this.  I guess the line number of the error will show the C function call?  I'm not sure if that's going to be harder for the user to figure out how to correct their code.  Probably it's fine, but worth checking how that looks to make sure you're not making obfuscated tracebacks.

Reply:
I will check the traceback. I added it here mostly to keep the error checking in the swig interface intact.

Reply:
BTW @rmjarvis. The python code always calls the C code through the functions here: https://github.com/LSSTDESC/CCL/blob/master/pyccl/pyutils.py#L67. These functions always pass the right size. So there is not a place to put these checks in the python code as you suggest. 

Given this, they may not even be needed. Thoughts?

Reply:
Also, the backtrace looks fine to me.

```python
In [1]: import pyccl                                                                                                                                                                                    

In [2]: pyccl.ccllib.growth_factor_vec?                                                                                                                                                                 
Signature: pyccl.ccllib.growth_factor_vec(cosmo, a, nout, status)
Docstring: growth_factor_vec(cosmology cosmo, double * a, int nout, int * status)
File:      ~/Desktop/CCL/pyccl/ccllib.py
Type:      function

In [3]: cosmo = ccl.Cosmology(Omega_c=0.27, Omega_b=0.045, h=0.67, sigma8=0.8, n_s=0.96)                                                                                                                
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-3-ffb1d6c5f82d> in <module>
----> 1 cosmo = ccl.Cosmology(Omega_c=0.27, Omega_b=0.045, h=0.67, sigma8=0.8, n_s=0.96)

NameError: name 'ccl' is not defined

In [4]: cosmo = pyccl.Cosmology(Omega_c=0.27, Omega_b=0.045, h=0.67, sigma8=0.8, n_s=0.96)                                                                                                              

In [5]: pyccl.ccllib.growth_factor_vec(cosmo, [0.0, 1.0, 2.0], 2, 0)                                                                                                                                    
---------------------------------------------------------------------------
CCLError                                  Traceback (most recent call last)
<ipython-input-5-72f9c8633d66> in <module>
----> 1 pyccl.ccllib.growth_factor_vec(cosmo, [0.0, 1.0, 2.0], 2, 0)

~/Desktop/CCL/pyccl/ccllib.py in growth_factor_vec(cosmo, a, nout, status)
    647 
    648     if numpy.shape(a) != (nout,):
--> 649         raise CCLError("Input shape for `a` must match `(nout,)`!")
    650 
    651 

CCLError: Input shape for `a` must match `(nout,)`!

In [6]:  
```

Note, as I described above, there are no higher level calls that are not wrapped in a `_vectorize_fn_*` function.

Reply:
Note also the nice function signature from switching the order of the argout size and pointer. :)

Reply:
Well, the main point was to avoid problems like [this](#425).  As long as the code doesn't let the user accidentally crash the interpreter, I don't much care how you go about it.

Although, now that I'm looking, I don't see the typo version as a unit test.  It should be.  You should check that the typo that used to be in the notebook raises an exception rather than crashing the interpreter.

Reply:
Sure I can add the unit test!